subsys/testsuite/ztest/include/zephyr/ztest_assert.h

```diff
@@ -476,6 +549,149 @@ static inline bool z_zassume(bool cond, const char *default_msg, const char *fil
  * @}
  */
 
+/**
+ * @defgroup ztest_expect Ztest expectation macros
+ * @ingroup ztest
+ *
+ * This module provides expectations when using Ztest.
+ *
+ * @{
+ */
+
+/**
+ * @brief Expect that @a cond is true, otherwise mark test as failed but continue its execution.
+ *
+ * @param cond Condition to check
+ * @param ... Optional message and variables to print if the expectation fails
+ */
+#define zexpect_true(cond, ...) zexpect(cond, #cond " is false", ##__VA_ARGS__)
+
+/**
+ * @brief Expect that @a cond is false, otherwise mark test as failed but continue its execution.
+ *
+ * @param cond Condition to check
+ * @param ... Optional message and variables to print if the expectation fails
+ */
+#define zexpect_false(cond, ...) zexpect(!(cond), #cond " is true", ##__VA_ARGS__)
+
+/**
+ * @brief Expect that @a cond is 0 (success), otherwise mark test as failed but continue its
+ * execution.
+ *
+ * @param cond Condition to check
+ * @param ... Optional message and variables to print if the expectation fails
+ */
+#define zexpect_ok(cond, ...) zexpect(!(cond), #cond " is non-zero", ##__VA_ARGS__)
+
+/**
+ * @brief Expect that @a ptr is NULL, otherwise mark test as failed but continue its execution.
+ *
+ * @param ptr Pointer to compare
+ * @param ... Optional message and variables to print if the expectation fails
+ */
+#define zexpect_is_null(ptr, ...) zexpect((ptr) == NULL, #ptr " is not NULL", ##__VA_ARGS__)
+
+/**
+ * @brief Expect that @a ptr is not NULL, otherwise mark test as failed but continue its execution.
+ *
+ * @param ptr Pointer to compare
+ * @param ... Optional message and variables to print if the expectation fails
+ */
+#define zexpect_not_null(ptr, ...) zexpect((ptr) != NULL, #ptr " is NULL", ##__VA_ARGS__)
+
+/**
+ * @brief Expect that @a a equals @a b, otherwise mark test as failed but continue its execution.
+ * expectation fails, the test will be marked as "skipped".
+ *
+ * @param a Value to compare
+ * @param b Value to compare
+ * @param ... Optional message and variables to print if the expectation fails
+ */
+#define zexpect_equal(a, b, ...) zexpect((a) == (b), #a " not equal to " #b, ##__VA_ARGS__)
+
+/**
+ * @brief Expect that @a a does not equal @a b, otherwise mark test as failed but continue its
+ * execution.
+ *
+ * @a a and @a b won't be converted and will be compared directly.
+ *
+ * @param a Value to compare
+ * @param b Value to compare
+ * @param ... Optional message and variables to print if the expectation fails
+ */
+#define zexpect_not_equal(a, b, ...) zexpect((a) != (b), #a " equal to " #b, ##__VA_ARGS__)
+
+/**
+ * @brief Expect that @a a equals @a b, otherwise mark test as failed but continue its execution.
+ *
+ * @a a and @a b will be converted to `void *` before comparing.
```

Comment:
Is this important for the caller to know?

Reply:
Probably: https://stackoverflow.com/q/916051

Reply:
I'm curious whether, on some of those platforms, it's even possible to convert pointers with different underlying representations to `void *` and have them be equal. But it seems like there isn't another reasonable way to compare pointers of different types. Unless you would expect the caller to do something different if you weren't casting to `void *`, I'd call this an implementation detail and leave it out. But up to you.